### PR TITLE
Added support for "large limits"

### PR DESCRIPTION
**Pull request by francofusco**
This PR proposes a new function to deal with shortest angle calculations in presence of "*large limits*", *i.e.*, when either `left_limit` or `right_limit` are outside the range `[ -pi , pi ]`.

### Motivating example

While using the `effort_controllers/JointGroupPositionController`, I experienced an issue similar to #2. In my specific case,  a call to `shortest_angular_distance_with_limits` was performed with:
```
from = 0.999507
to = 1.0
left_limit = -62.831853 (-20*pi)
right_limit = 62.831853 (20*pi)
```
the function returned `false`, and the angle was evaluated to `-6.282692`. This is not really an error, since it is written in the documentation that limits are supposed to be within `-pi` and `+pi` and in addition the function returned `false`. However, the validity of the limits is not checked inside `shortest_angular_distance_with_limits`, and thus it's user's responsibility to make sure that such calls are not performed.

I however think that having `shortest_angular_distance_with_limits` to work with bounds in `[ -pi , pi ]` is a strong limitation. I am aware that having limits in a range such as `[ -20*pi , 20*pi ]` for a revolute joint is uncommon and that a continuous joint would solve the issue without posing many limitations. Nonetheless, I think it would be more robust to explicitly add support for "large limits".

### Proposed solution

I implemented the function `shortest_angular_distance_with_large_limits`, which is supposed to work with bounds larger than `pi`.

The main difference with `shortest_angular_distance_with_limits` is that it does not allow "*reversed bounds*" such as `[ 0.75*pi , -0.75*pi ]`. Reverse bounds make sense in the unit circle where the rotation `x+2*pi` is considered exactly the same as `x`. However, I believe the same should not be true when considering larger bounds, and my function thus requires `left_limit < right_limit`.

The goal of the function is to return the smallest angle `x` such that `fmod(from+x, 2*pi)` equals `fmod(to, 2*pi)` also ensuring that `left_limit <= from+x <= right_limit`. Note that `to` does not need to be in the valid interval. As an example, `shortest_angular_distance_with_limits(0, 0.5*pi, -2*pi, 2*pi, sa)` will return `true` with `sa=0.5*pi`, while `shortest_angular_distance_with_limits(0, 0.5*pi, -2*pi, 0.1*pi, sa)` will return `true` with `sa=-1.5*pi` (since `0.5*pi` is larger than `right_limit`).

I propose to let `shortest_angular_distance_with_limits` automatically call the new function whenever a bound is larger than `pi` (in magnitude). Since I know this might break back-compatibility in many ways, I am open to alternative solutions!

### Minor changes

- Added few tests in angles/test/utest.cpp
- Added the `pthread` flag in angles/test/CMakeLists.txt - I had linker errors while building the tests, and followed what reported in [gtest's wiki page](http://wiki.ros.org/gtest#Building_and_running_tests)

### CODE REVIEW

File: angles/test/CMakeLists.txt [hunk 1]
```diff
@@ -1,4 +1,5 @@
 if (CATKIN_ENABLE_TESTING)
+  set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -pthread")
```
Comment (tfoote) [inline angles/test/CMakeLists.txt]: This seems unrelated.

File: angles/test/utest.cpp [hunk 2]
```diff
@@ -160,7 +195,7 @@ TEST(Angles, normalize_angle)
  EXPECT_NEAR(M_PI/2, normalize_angle(5*M_PI/2), epsilon);
  EXPECT_NEAR(M_PI/2, normalize_angle(9*M_PI/2), epsilon);
  EXPECT_NEAR(M_PI/2, normalize_angle(-3*M_PI/2), epsilon);
- 
+
```
Comment (tfoote) [inline angles/test/utest.cpp]: We try not to make unnecessary whitespace only changes if possible to avoid future merge conflicts.

File: angles/include/angles/angles.h [hunk 3]
```diff
@@ -215,6 +290,13 @@ namespace angles
   static inline bool shortest_angular_distance_with_limits(double from, double to, double left_limit, double right_limit, double &shortest_angle)
   {
 
+    // If either left_limit or right_limit is outside the "standard" unit circle,
+    // call shortest_angular_distance_with_large_limits instead
+    if(std::fabs(left_limit) > M_PI || std::fabs(right_limit) > M_PI)
+    {
+      return shortest_angular_distance_with_large_limits(from, to, left_limit, right_limit, shortest_angle);
```
Comment (tfoote) [inline angles/include/angles/angles.h]: With the description and parallel implementation I was surprised to see this here. I'm not sure that it makes sense for behavior to change if I slowly increase the limit value past a magic number. In particular this I think will yield quite different answers if you are doing operations near the back of the unit circle and you're leveraging the wrap around the back. Suddenly your check will return false if you shift your interval across the back of the unit circle.

I think that it might make more sense to keep them separate and clarify that the existing one is computing on a unit circle and that all input values would be normalized onto the [-M_PI, M_PI) range.

And the logic for a multiple turn joint is kept separate in your new computation.

File: angles/test/CMakeLists.txt (hunk 1 above)
Comment (francofusco) [reply angles/test/CMakeLists.txt]: Yes, this is indeed unrelated to the issue I wanted to solve. However, without it I was not able to build and run the tests. I felt it was alright to add this change here, but I can make a separate PR in case you think it would be more appropriate.

File: angles/test/utest.cpp (hunk 2 above)
Comment (francofusco) [reply angles/test/utest.cpp]: Sorry, I did not notice this change. I will restore it back to the original state.

File: angles/include/angles/angles.h (hunk 3 above)
Comment (francofusco) [reply angles/include/angles/angles.h]: I understand. I think my mistake was in assuming that `shortest_angular_distance_with_limits` would be used with `left_limit` and `right_limit` strictly inside the `[-pi,pi)` range. I will remove this change and update the description of `shortest_angular_distance_with_large_limits`.

File: angles/test/utest.cpp [hunk 4]
```diff
@@ -130,7 +165,7 @@ TEST(Angles, normalize_angle_positive)
  EXPECT_NEAR(M_PI/2, normalize_angle_positive(5*M_PI/2), epsilon);
  EXPECT_NEAR(M_PI/2, normalize_angle_positive(9*M_PI/2), epsilon);
  EXPECT_NEAR(M_PI/2, normalize_angle_positive(-3*M_PI/2), epsilon);
-
+
```
Comment (tfoote) [inline angles/test/utest.cpp]: I think you added the whitespace back in the wrong spot.

File: angles/test/CMakeLists.txt (hunk 1 above)
Comment (tfoote) [reply angles/test/CMakeLists.txt]: Yes, please separate this. There's something else going on then in your workspace as this is passing CI w/o this change. Hard coding `-pthread` isn't a portable solution either.

Comment (francofusco) [reply angles/test/CMakeLists.txt]: Ok, I will remove it and check what's wrong with my setup.

File: angles/test/utest.cpp (hunk 4 above)
Comment (francofusco) [reply angles/test/utest.cpp]: Yes, my bad.

**Comment by francofusco** [comment]
I followed the reviews and now the changes are limited to the new function and its related tests. I finally added the same function also in the python API for consistency and included the relevant tests.

**Comment by francofusco** [comment]
Thanks to you for the review!
